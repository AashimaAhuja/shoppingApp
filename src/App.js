import React, { useState } from "react";
import CardList from "./components/CardList";
import Header from "./components/Header";
import ProductsTitle from "./components/ProductsTitle";
import { getProductList } from "./services/productService";
import "./App.css";

function App({
  testIsLoading,
  testShowSoldItems,
  testVisibleProducts
}) {
  const [productList, setProductList] = useState([]);
  const [visibleProducts, setVisibleProducts] = useState(testVisibleProducts ?? []);
  const [showSoldItems, setShowSoldItems] = useState(testShowSoldItems ?? true);
  const [isLoading, setIsLoading] = useState(testIsLoading ?? true);

  React.useEffect(() => {
    getProductList()
      .then(setProductList)
      .catch(console.log("Error occured"))
      .finally(setIsLoading(false));
  }, []);

  React.useEffect(() => {
    if (showSoldItems) {
      setVisibleProducts(productList);
    } else {
      let filteredProducts = visibleProducts.filter((product) => !product.sold);
      setVisibleProducts(filteredProducts);
    }
  }, [productList, showSoldItems, visibleProducts]);

  const onToggleSoldItems = React.useCallback(() => {
    setShowSoldItems((showSoldItems) => !showSoldItems);
  }, []);

  const onProductLike = React.useCallback(
    (id) => {
      let likedProduct = visibleProducts.find((product) => product.id === id);
      likedProduct.isLiked = !likedProduct.isLiked;
      setVisibleProducts([...visibleProducts]);
    },
    [visibleProducts]
  );

  return (
    <div className="App">
      <Header visibleProducts={visibleProducts} />
      {isLoading && <div className='loading'>Loading</div>}
      {!isLoading && visibleProducts.length > 0 && (
        <main>
          <ProductsTitle
            productsCount={visibleProducts.length}
            onToggleSoldItems={onToggleSoldItems}
            showSoldItems={showSoldItems}
          />
          <CardList products={visibleProducts} onProductLike={onProductLike} />
        </main>
      )}
    </div>
  );
}

export default App;
