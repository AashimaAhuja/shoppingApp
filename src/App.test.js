import { shallow } from 'enzyme';
import React from 'react';
import App from './App';

const defaultProps = {
  testVisibleProducts: [
    {
      isLiked: true,
      id: 1,
      name: 'prod 1'
    },
    {
      isLiked: true,
      id: 2,
      name: 'prod 1'
    },
    {
      isLiked: true,
      id: 3,
      name: 'prod 1'
    },
    {
      isLiked: true,
      id: 4,
      name: 'prod 1'
    },
    {
      isLiked: false,
      id: 5,
      name: 'prod 1'
    },
  ],
  testIsLoading: false,
  testShowSoldItems: false
}

describe('<App/> Tests', () => {
  let useEffect, useCallback;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f())
  }

  const mockUseCallback = () => {
    useCallback.mockImplementationOnce(f => f())
  }

  const mockUseCallbackWithId = (id) => {
    useCallback.mockImplementationOnce(f => f(id))
  }

  beforeAll(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    useCallback = jest.spyOn(React, 'useCallback');
    mockUseEffect();
    mockUseEffect();
    shallow(<App />)
  })

  it('should render App Component', () => {
    const AppComponent = shallow(<App />);
    expect(AppComponent).toMatchSnapshot();
  });

  it('should render App Component with default props', () => {
    mockUseEffect();
    mockUseEffect();
    mockUseCallback();
    mockUseCallbackWithId(1);
    const AppComponent = shallow(<App {...defaultProps} />)
    expect(AppComponent).toMatchSnapshot();
  });

  it('should render App Component with is Loading true', () => {
    const wrapper = shallow(<App testIsLoading={true} />)
    expect(wrapper).toMatchSnapshot();
    const loadingDiv = wrapper.find('.loading');
    expect(loadingDiv.text()).toEqual('Loading');
  });
})
