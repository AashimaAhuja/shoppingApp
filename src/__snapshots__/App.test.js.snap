// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> Tests should render App Component 1`] = `
<div
  className="App"
>
  <Header
    visibleProducts={Array []}
  />
  <div
    className="loading"
  >
    Loading
  </div>
</div>
`;

exports[`<App/> Tests should render App Component with default props 1`] = `
<div
  className="App"
>
  <Header
    visibleProducts={
      Array [
        Object {
          "id": 1,
          "isLiked": false,
          "name": "prod 1",
        },
        Object {
          "id": 2,
          "isLiked": true,
          "name": "prod 1",
        },
        Object {
          "id": 3,
          "isLiked": true,
          "name": "prod 1",
        },
        Object {
          "id": 4,
          "isLiked": true,
          "name": "prod 1",
        },
        Object {
          "id": 5,
          "isLiked": false,
          "name": "prod 1",
        },
      ]
    }
  />
  <main>
    <ProductsTitle
      productsCount={5}
      showSoldItems={true}
    />
    <CardList
      products={
        Array [
          Object {
            "id": 1,
            "isLiked": false,
            "name": "prod 1",
          },
          Object {
            "id": 2,
            "isLiked": true,
            "name": "prod 1",
          },
          Object {
            "id": 3,
            "isLiked": true,
            "name": "prod 1",
          },
          Object {
            "id": 4,
            "isLiked": true,
            "name": "prod 1",
          },
          Object {
            "id": 5,
            "isLiked": false,
            "name": "prod 1",
          },
        ]
      }
    />
  </main>
</div>
`;

exports[`<App/> Tests should render App Component with is Loading true 1`] = `
<div
  className="App"
>
  <Header
    visibleProducts={Array []}
  />
  <div
    className="loading"
  >
    Loading
  </div>
</div>
`;
